File: 303. Range Sum Query - Immutable.py
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.totalnums = nums
        for i in xrange(1, len(self.totalnums)):
            self.totalnums[i] += self.totalnums[i - 1]
        

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        if i == 0:
            return self.totalnums[j]
        return self.totalnums[j] - self.totalnums[i - 1]


# Your NumArray object will be instantiated and called as such:
# numArray = NumArray(nums)
# numArray.sumRange(0, 1)
# numArray.sumRange(1, 2)

File: 303. Range Sum Query - Immutable.cpp
class NumArray {
private:
    vector<int> totalnums;
public:
    NumArray(vector<int> &nums) {
        totalnums = nums;
        for (int i = 1; i < totalnums.size(); i++)
            totalnums[i] += totalnums[i-1];
    }

    int sumRange(int i, int j) {
        return i == 0 ? totalnums[j] : totalnums[j] - totalnums[i-1];
    }
};


// Your NumArray object will be instantiated and called as such:
// NumArray numArray(nums);
// numArray.sumRange(0, 1);
// numArray.sumRange(1, 2);
